src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -121,13 +130,29 @@ protected static JsonConverter<TValue> GetValueConverter(ref WriteStack state)
                         // Read method would have thrown if otherwise.
                         Debug.Assert(reader.TokenType == JsonTokenType.PropertyName);
 
-                        state.Current.JsonPropertyNameAsString = reader.GetString();
+                        // Copy key name for JSON Path support in case of error.
+                        string unescapedPropertyNameAsString;
+
+                        TKey key;
+                        if (IsStringKey)
+                        {
+                            object keyAsObject = reader.GetString()!;
+                            key = (TKey)keyAsObject;
+                            unescapedPropertyNameAsString = (string)keyAsObject;
+                        }
+                        else
+                        {
+                            JsonConverter<TKey> keyConverter = GetKeyConverter(options);
+                            key = keyConverter.ReadWithQuotes(ref reader);
+                            unescapedPropertyNameAsString = reader.GetString()!;
+                        }
 
+                        state.Current.JsonPropertyNameAsString = unescapedPropertyNameAsString;
```

Comment:
Extracting this into an instance method was causing a perf. decrease; I tried using `AggresiveInlining` but no luck.

Reply:
For now I moved this code into a local method since that seems to regress less the performance.
There might be a reason why the method doesn't get inlined (even with AggressiveInlining) but for now these are the perf results:

worse: 8, geomean: 1.029
total diff: 8

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadJson<Dictionary<String, String>>.Deseri |      1.05 |         21061.70 |         22172.59 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableSortedDictionary<String,  |      1.04 |         70776.79 |         73581.40 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Hashtable>.DeserializeFromStream   |      1.04 |         62335.99 |         64625.97 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Dictionary<String, String>>.Deseri |      1.03 |         19661.33 |         20254.28 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableSortedDictionary<String,  |      1.03 |         69778.78 |         71658.05 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableDictionary<String, String |      1.03 |         46107.37 |         47298.85 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableDictionary<String, String |      1.02 |         44138.14 |         44838.34 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Hashtable>.DeserializeFromUtf8Byte |      1.01 |         62344.90 |         62836.07 |         |

No Faster results for the provided threshold = 0% and noise filter = 0.3ns.